### PR TITLE
DB-10562 Refactor SpliceTransactionResourceImpl.getMarshallTransaction 

### PR DESCRIPTION
**Pull request by arnaud-splice**

### CODE REVIEW

**Review by martinrupp — APPROVED** [review]

File: splice_machine/src/main/java/com/splicemachine/derby/ddl/DDLUtils.java [hunk 1]
```diff
@@ -376,36 +367,45 @@ public static void preAlterStats(DDLMessage.DDLChange change, DataDictionary dd,
...
+        DDLMessage.DropSchema dropSchema = change.getDropSchema();
+        UUID schemaId = ProtoUtil.getDerbyUUID(dropSchema.getSchemaUUID());
+        try {
+            try (SpliceTransactionResourceImpl transactionResource = new SpliceTransactionResourceImpl()) {
+                TxnView txn = DDLUtils.getLazyTransaction(change.getTxnId());
+                transactionResource.marshallTransaction(txn);
+                SchemaDescriptor sd = dd.getSchemaDescriptor(schemaId, transactionResource.getLcc().getTransactionExecute());
+                if(sd==null) // Schema Descriptor transaction never committed
+                    return;
+                dm.invalidateFor(sd,DependencyManager.DROP_SCHEMA,transactionResource.getLcc());
```
Comment (hatyo) [inline splice_machine/src/main/java/com/splicemachine/derby/ddl/DDLUtils.java:379]: Looks like an additional logic besides refactoring right? Could you explain why this is necessary?

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/index/IndexTransformer.java [hunk 2]
```diff
@@ -707,18 +707,12 @@ public int getMaxBaseColumnPosition() {
     }
 
     private LanguageConnectionContext getLcc(DDLMessage.TentativeIndex tentativeIndex) throws StandardException {
-        boolean prepared = false;
-        SpliceTransactionResourceImpl transactionResource = null;
-        try {
-            TxnView txn = DDLUtils.getLazyTransaction(tentativeIndex.getTxnId());
-            transactionResource = new SpliceTransactionResourceImpl();
-            prepared = transactionResource.marshallTransaction(txn);
+        TxnView txn = DDLUtils.getLazyTransaction(tentativeIndex.getTxnId());
```
Comment (arnaud-splice) [inline splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/index/IndexTransformer.java:710]: Returning the LCC here seems wrong. Rework that class a bit to not have to return a LCC whose contextmanager got closed

File: splice_machine/src/main/java/com/splicemachine/derby/jdbc/SpliceTransactionResourceImpl.java [hunk 3]
```diff
@@ -79,17 +80,17 @@ public SpliceTransactionResourceImpl(String url,Properties info) throws SQLExcep
...
-        return this.marshallTransaction(txn, propertyCache, null, null, null);
+    public void marshallTransaction(TxnView txn, ManagedCache<String, Optional<String>> propertyCache) throws StandardException, SQLException {
+        this.marshallTransaction(txn, propertyCache, null, null, null);
     }
 
-    public boolean marshallTransaction(TxnView txn, ManagedCache<String, Optional<String>> propertyCache,
+    public void marshallTransaction(TxnView txn, ManagedCache<String, Optional<String>> propertyCache,
                                        TransactionController reuseTC, String localUserName, Integer sessionNumber) throws StandardException, SQLException{
-        boolean updated = false;
+        assert !updated;
```
Comment (hatyo) [inline splice_machine/src/main/java/com/splicemachine/derby/jdbc/SpliceTransactionResourceImpl.java:93]: the method sets the `updated` to true and is public, so it is possible that it is called twice from the outside (which was permissible in the previous version)

I think it is better to return immediately and maybe issue a log.WARN("unexpected") or even throw an exception.

I would also add more documentation about the intended usage of this method, and mark the class `SpliceTransactionResourceImpl` as thread-unsafe.

**Review by hatyo — CHANGES_REQUESTED** [review]
It is really nice that you replaced `try ... catch` with `try-with-resource` where applicable. But I am afraid this PR is doing a bit more besides refactoring. Please have a look at my comments.
Excellent work!

File: splice_machine/src/main/java/com/splicemachine/derby/jdbc/SpliceTransactionResourceImpl.java (hunk 3 above)
Comment (arnaud-splice) [reply splice_machine/src/main/java/com/splicemachine/derby/jdbc/SpliceTransactionResourceImpl.java:93]: Done

File: splice_machine/src/main/java/com/splicemachine/derby/ddl/DDLUtils.java (hunk 1 above)
Comment (arnaud-splice) [reply splice_machine/src/main/java/com/splicemachine/derby/ddl/DDLUtils.java:379]: This is true. We found out with Daniel that the code in `preDropSchema` did not invalidate potential dependencies to that schema. In a pre multidb world, it probably does not matter, but we should do in `preDropSchema` what we do in `preDropTable` anyway

Comment (hatyo) [reply splice_machine/src/main/java/com/splicemachine/derby/ddl/DDLUtils.java:379]: OK, thanks for the clarification.

File: splice_machine/src/main/java/com/splicemachine/derby/jdbc/SpliceTransactionResourceImpl.java (hunk 3 above)
Comment (hatyo) [reply splice_machine/src/main/java/com/splicemachine/derby/jdbc/SpliceTransactionResourceImpl.java]: Great

**Review by hatyo — APPROVED** [review]

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/index/IndexTransformer.java (hunk 2 above)
Comment (jyuanca) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/index/IndexTransformer.java:714]: lcc is only used in getExecutableIndexExpression(). You can refactor to marshall transactions there only when it is being used

**Review by jyuanca — CHANGES_REQUESTED** [review]

File: splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/index/IndexTransformer.java (hunk 2 above)
Comment (arnaud-splice) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/index/IndexTransformer.java:714]: There is a dedicated jira for this specific issue: DB-10485. Zhen is currently working on it, so I would prefer not modify this further.

**Review by jyuanca — APPROVED** [review]

**Pull request merged**